tests: net: socket: tcp: Add test for big block data transfer

ssharks opened the pull request:
Add tests to reproduce the issues described in:
#44018
#44012

It send a data chunk from server to client over the loopback interface. It repeats the same, while introducing packet loss at the loopback driver.

rlubos commented on drivers/net/Kconfig:
```diff
@@ -183,6 +183,12 @@ menuconfig NET_LOOPBACK
 
+config NET_LOOPBACK_PACKET_DROP
+	bool "Controlable packet drop"
+	default n
```
No need for `default n`, it's implicit.

rlubos commented on drivers/net/loopback.c:
```diff
@@ -60,6 +60,15 @@ static void loopback_init(struct net_if *iface)
+static float loopback_packet_drop_ratio = 0.0f;
+static float loopback_packet_drop_state = 0.0f;
+
+void loopback_set_packet_drop_ratio(float ratio) {
```
The API function should be exposed in a header file. I see there's no `net/loopback.h` yet, I think it should be added.

rlubos commented on drivers/net/loopback.c:
```diff
@@ -102,11 +111,27 @@ static int loopback_send(const struct device *dev, struct net_pkt *pkt)
+		}
+#ifdef CONFIG_NET_LOOPBACK_PACKET_DROP
+	} else {
+		// administrate we dropped a packet
```
Please use C-style comments.

ssharks replied:
Done

rlubos commented on drivers/net/loopback.c:
```diff
@@ -102,11 +111,27 @@ static int loopback_send(const struct device *dev, struct net_pkt *pkt)
+	/* Drop packets based on the loopback_packet_drop_ratio
+	a ratio of 0.2 will drop one every 5 packets */
+	loopback_packet_drop_state += loopback_packet_drop_ratio;
+	if (loopback_packet_drop_state < 1.0f) {
```
If you reverse the logic here (`if (loopback_packet_drop_state  >= 1.0f) {`) you could have a single, cocise `#ifdef` block. It could end up with `goto out;` so you won't even need to modify existing code.

ssharks replied:
Thanks for the suggestion, done

rlubos commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -234,15 +234,59 @@ void test_v6_send_recv(void)
+/* Test the stack behavior with a resonable sized block data, be sure to have multiple packets */
+#define TEST_BLOCK_SIZE 60000
+
+static uint8_t testBlockTcp[TEST_BLOCK_SIZE];
+static uint8_t testBlockTcpRx[TEST_BLOCK_SIZE + 10];
```
This will greatly increase the RAM requirements of the test suite. Perhaps it'd be better to use smaller buffers (1k for example) and just re-send it several times?

rlubos replied:
Also, please use snake_case for Zephyr code.

rlubos commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -234,15 +234,59 @@ void test_v6_send_recv(void)
+		total_received += recved;
+		iteration++;
+	}
+	zassert_equal(memcmp(testBlockTcpRx, testBlockTcp, sizeof(testBlockTcp)),
```
There's `zassert_mem_equal()` for this such checks.

rlubos commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -252,26 +296,90 @@ void test_v4_sendto_recvfrom(void)
-
-	test_accept(s_sock, &new_sock, &addr, &addrlen);
-	zassert_equal(addrlen, sizeof(struct sockaddr_in), "wrong addrlen");
+	(void)k_thread_create(&tcp_server_thread_data, tcp_server_stack_area,
+	                                 K_THREAD_STACK_SIZEOF(tcp_server_stack_area),
+	                                 tcp_server_block_thread,
+	                                 &s_sock, NULL, NULL,
+	                                 k_thread_priority_get( k_current_get()), 0, K_NO_WAIT);
```
Please fix the alignment (align with the opening bracket).

rlubos commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -252,26 +296,90 @@ void test_v4_sendto_recvfrom(void)
-	test_recvfrom(new_sock, MSG_PEEK, &addr, &addrlen);
-	zassert_equal(addrlen, sizeof(struct sockaddr_in), "wrong addrlen");
+	test_connect(c_sock, (struct sockaddr *)&s_saddr, sizeof(s_saddr));
+	LOG_ERR("Test this log message");
```
I don't think it's needed in the test.

ssharks replied:
Indeed obsolete code

rlubos commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -234,15 +234,59 @@ void test_v6_send_recv(void)
+	test_close(new_sock);
+}
+
+void test_v4_send_recv_block(void)
```
I wonder about the name here, `block` suggests that the test is about the blocking socket, while it's not essential here. Maybe rename it to `test_v4_send_recv_large`?

ssharks replied:
Good suggestion, done

rlubos commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -1075,7 +1183,10 @@ void test_main(void)
 		ztest_unit_test(test_v6_msg_waitall),
-		ztest_user_unit_test(test_socket_permission)
+		ztest_user_unit_test(test_socket_permission),
+		ztest_unit_test(test_v4_send_recv_block),
+		ztest_unit_test_setup_teardown(test_v4_send_recv_block,
```
The test is added twice, is that intended?

ssharks replied:
Yes, the first one is without packet loss, the second one with

ssharks commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -234,15 +234,59 @@ void test_v6_send_recv(void)
-	socklen_t addrlen = sizeof(addr);
+
+	/* initialize the test block with easily debuggable data */
+	for (uint32_t i=0; i<TEST_BLOCK_SIZE; i++) {
```
Done

jukkar commented on drivers/net/Kconfig:
```diff
@@ -183,6 +183,11 @@ menuconfig NET_LOOPBACK
+config NET_LOOPBACK_SIMULATE_PACKET_DROP
+	bool "Controlable packet drop"
+	help
+	  Enable interface to have a controlable packet drop rate
```
It would be nice to emphasize that this is only used by testing and should not be enabled by end user.

jukkar commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -234,6 +235,114 @@ void test_v6_send_recv(void)
+static uint8_t testBlockTcp[TEST_BLOCK_SIZE];
+static uint8_t testBlockTcpRx[TEST_BLOCK_SIZE + 10];
+
+#define TCP_SERVER_STACK_SIZE 4096
```
This and the TEST_BLOCK_SIZE will most definitely mean that the unit test cannot be run on all the devices it is currently running by twister. I just wonder could we make these numbers a bit smaller?

ssharks replied:
I'll give it a try, the challenge that is needs to be a certain size to trigger the issue...

rlubos replied:
Yes, but you can just keep sending the same buffer over and over, the data doesn't need to be unique (of course it would still be good to verify data integrity after each iteration).

ssharks replied:
I have significantly reduced this, the only thing left is 2048 bytes of stack size, the issues still reproduces without the fixes

rlubos commented on include/net/loopback.h:
```diff
@@ -0,0 +1,31 @@
+ */
+
+/*
+ * Copyright (c) 2017 Intel Corporation
```
2022. Please add yourself in the copyright.

ssharks replied:
You're totally right, lazy copy paste action

rlubos commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -245,16 +354,16 @@ void test_v4_sendto_recvfrom(void)
 	socklen_t addrlen = sizeof(addr);
 
 	prepare_sock_tcp_v4(CONFIG_NET_CONFIG_MY_IPV4_ADDR, ANY_PORT,
-			    &c_sock, &c_saddr);
```
Please don't introduce unnecessary changes.

rlubos commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -245,16 +354,16 @@ void test_v4_sendto_recvfrom(void)
-			    &c_sock, &c_saddr);
+				&c_sock, &c_saddr);
 	prepare_sock_tcp_v4(CONFIG_NET_CONFIG_MY_IPV4_ADDR, SERVER_PORT,
-			    &s_sock, &s_saddr);
```
Also here.

rlubos commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -245,16 +354,16 @@ void test_v4_sendto_recvfrom(void)
 	test_connect(c_sock, (struct sockaddr *)&s_saddr, sizeof(s_saddr));
 	test_sendto(c_sock, TEST_STR_SMALL, strlen(TEST_STR_SMALL), 0,
-		    (struct sockaddr *)&s_saddr, sizeof(s_saddr));
+			(struct sockaddr *)&s_saddr, sizeof(s_saddr));
```
And here.

ssharks replied:
Same here, reverted

rlubos commented:
Could you also split the changes in the loopback interface and the actual test into separate commits?

ssharks committed:
drivers: net: loopback: Add interface to simulate packet drop

To allow for high level robustness tests on protocols, add an interface
to control the packet drop rate. A rate of 0 means no packet dropped, a
rate of 1 means all packets being dropped.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks committed:
tests: net: socket: tcp: Add test for big block data transfer

A reproducible case to see what happens when the TCP stack runs out
of buffers. It transfers a block, bigger then the number of buffers
available.
Also test by introducing packet loss in the loopback driver.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks commented:
Thanks for your review input, this helps to get the test quality up
